src/GLX/glvnd_genentry.c

```diff
@@ -49,299 +49,157 @@
 
 #if defined(__GNUC__) && USE_ASM
 
-/// The maximum number of entrypoints that we can generate.
-#define GENERATED_ENTRYPOINT_MAX 4096
+#define _U_STRINGIFY(x) #x
+#define U_STRINGIFY(x) _U_STRINGIFY(x)
 
-/// The size of each generated entrypoint.
-static const int STUB_ENTRY_SIZE = 32;
+#define GLX_STUBS_COUNT
+#include "g_glx_dispatch_stub_list.h"
+
+static GLVNDentrypointStub entrypointFunctions[GENERATED_ENTRYPOINT_MAX];
+static char *entrypointNames[GENERATED_ENTRYPOINT_MAX] = {};
+static int entrypointCount = 0;
+
+extern char glx_entrypoint_start[];
+extern char glx_entrypoint_end[];
 
 #if defined(USE_X86_ASM)
-/// A template used to generate an entrypoint.
-static unsigned char STUB_TEMPLATE[] =
-{
-    0xe9, 0x78, 0x56, 0x34, 0x12, // jmp 0x12345678
-};
 
-static const int DISPATCH_FUNC_OFFSET = 1;
-static const int DISPATCH_FUNC_OFFSET_REL = 5;
+#define STUB_SIZE 32
+#define STUB_ASM_ARCH(slot) \
+    "push %ebx\n" \
+    "call 1f\n" \
+    "1:\n" \
+    "popl %ebx\n" \
+    "addl $_GLOBAL_OFFSET_TABLE_+[.-1b], %ebx\n" \
+    "movl entrypointFunctions@GOT(%ebx), %eax\n" \
+    "pop %ebx\n" \
+    "jmp *(4 * " slot ")(%eax)\n"
 
 #elif defined(USE_X86_64_ASM)
-// For x86_64, the offset from the entrypoint to the dispatch function might be
-// more than 2^31, and there's no JMP instruction that takes a 64-bit offset.
-// Note that the same stub also works for an x32 build. In that case, though, a
-// pointer is only 32 bits, so we have to make sure we expand it a 64-bit value
-// when we patch it in SetDispatchFuncPointer.
-static unsigned char STUB_TEMPLATE[] =
-{
-    0x48, 0xb8, 0xbd, 0xac, 0xcd, 0xab, 0x78, 0x56, 0x34, 0x12, // movabs 0x12345678abcdacbd,%rax
-    0xff, 0xe0, // jmp *%rax
-};
 
-static const int DISPATCH_FUNC_OFFSET = 2;
+#define STUB_SIZE 16
+#define STUB_ASM_ARCH(slot) \
+    "movq entrypointFunctions@GOTPCREL(%rip), %rax\n\t" \
+    "jmp *(8 * " slot ")(%rax)\n"
 
 #elif defined(USE_ARMV7_ASM)
-// Thumb bytecode
-static const uint16_t STUB_TEMPLATE[] =
-{
-    // ldr ip, 1f
-    0xf8df, 0xc004,
-    // bx ip
-    0x4760,
-    // nop
-    0xbf00,
-    // Offset that needs to be patched
-    // 1:
-    0x0000, 0x0000,
-};
-
-static const int DISPATCH_FUNC_OFFSET = 8;
+
+#define STUB_SIZE 32
+#define STUB_ASM_ARCH(slot) \
+    "ldr ip, 1f\n" \
+    "12:\n" \
+    "add ip, pc\n" \
+    "push { r0 }\n" \
+    "ldr r0, 1f+4\n" \
+    "add ip, r0\n" \
+    "pop { r0 }\n" \
+    "ldr ip, [ip]\n" \
+    "bx ip\n" \
+    "1:\n" \
+    ".word entrypointFunctions - (12b + 8)\n" \
+    ".word " slot " * 4\n"
 
 #elif defined(USE_AARCH64_ASM)
 
-static const uint32_t STUB_TEMPLATE[] =
-{
-    // ldr x16, 1f
-    0x58000070,
-    // br x16
-    0xd61f0200,
-    // nop
-    0xd503201f,
-    // Offset that needs to be patched
-    // 1:
-    0x00000000, 0x00000000,
-};
-
-static const int DISPATCH_FUNC_OFFSET = 12;
+#define STUB_SIZE 16
+#define STUB_ASM_ARCH(slot) \
+    "adrp x16, entrypointFunctions + " slot "*8\n" \
```

Comment:
Does the assembler convert this into:
```adrp x16, label
add x16, slot * 8```

I ask as I believe `adrp` will place the page-aligned address into `xd`. On that note, I think entrypointFunctions would also needs to be guaranteed to be 4k aligned.

Reply:
It doesn't generate a separate `add` instruction, but as far as I understand, it's not needed.

The `adrp` instruction will load all but the low-order 12 bits of the address into x16, and then the `ldr` instruction takes the remaining offset as an immediate value.

For both instructions, it's loading the address `(entrypointFunctions + slot * 8)`, not just `(entrypointFunctions)` by itself.

Reply:
Ah, I wasn't paying attention to the offset passed to ldr. Looks good then 👍